pennylane/math/quantum.py

```diff
@@ -720,6 +720,61 @@ def _compute_mutual_info(
     return vn_entropy_1 + vn_entropy_2 - vn_entropy_12
 
 
+def expectation_value(operator_matrix, state_vector, check_state=False, c_dtype="complex128"):
+    r"""Compute the expectation value of an operator with respect to a pure state.
+
+    The expectation value of an operator :math:`A` for a pure state given by a state vector :math:`\ket{\psi}`
+    is defined as
+
+    .. math::
+        \langle A \rangle_\psi = \bra{\psi} A \ket{\psi}
+
+    Args:
+        operator_matrix (tensor_like):  operator matrix with shape ``(2**N, 2**N)`` or ``(batch_dim, 2**N, 2**N)``.
+        state_vector (tensor_like): state vector with shape ``(2**N)`` or ``(batch_dim, 2**N)``.
+        check_state (bool): If True, the function will check the validity of the state vector
+            via its shape and the norm
+        c_dtype (str): Complex floating point precision type.
+
+    Returns:
+        float: Expectation value of the operator for the state vector.
+
+    **Example**
+
+    An operator matrix and a state vector can be used as arguments to obtain the expectation value, e.g.:
+
+    >>> state_vector = np.array([0, 1])
+    >>> operator_matrix = np.array([[0,1],[1,0]])
+    >>> qml.math.expectation_value(operator_matrix, state_vector)
+    0.0

```

Comment:
A little more explanatory text above the example might be useful. Perhaps something along the lines of "We define a state vector describing a single qubit in the excited state, and find the expectation value of :math:`\sigma_x` on the state:"

I would include a two-qubit example as well. And I don't know if this fits within the intended use, so please do disregard if it doesn't, but it might be nice to also include an example using a matrix from a PL operator. These could be combined into a single example, something like:

```
>>> state_vector = [1/np.sqrt(2), 0, 1/np.sqrt(2), 0]
>>> operator_matrix = qml.PauliZ(0).matrix(wire_order=[0,1])

>>> qml.math.expectation_value(operator_matrix, state_vector)
tensor(-2.23711432e-17+0.j, requires_grad=True)
```

Here again I think its clearest if we describe the state we define, and the operator, and in particular highlight using the `wire_order` argument to ensure a two-qubit matrix.

Reply:
Good idea! 

I'll use this example and add some tests for $N=2$.